File: backend/app/services/ocr_service.py
import base64
import httpx
import json
import io
from typing import Optional
from app.config import settings

# Cost optimization constants
# GPT-4o-mini vision: ~$0.00015 per image (low detail mode)
# We use low detail to minimize cost — enough for text extraction
IMAGE_DETAIL = "low"  
MAX_TOKENS = 400  # Drug list extraction needs very few tokens
MAX_PAGES = 2     # Only process first 2 pages — prescriptions 
                  # are usually 1 page

SYSTEM_PROMPT = """You are a prescription parser. 
Extract ONLY medication/drug names from the prescription image.
Return a JSON object with exactly this structure:
{
  "drugs": ["drug1", "drug2"],
  "confidence": "high" | "medium" | "low",
  "notes": "brief note if any"
}
Rules:
- Return generic names in lowercase when possible
- Include brand names if generic is unclear
- Do NOT include dosages, frequencies, or instructions
- If no drugs found, return {"drugs": [], "confidence": "low"}
- Return ONLY valid JSON, no other text"""

async def pdf_to_images_base64(
    pdf_bytes: bytes
) -> list[str]:
    """
    Convert PDF pages to base64 images using pypdf + pillow.
    Returns list of base64 encoded PNG strings.
    Cost optimization: max MAX_PAGES pages, 
    resize to 800px width to reduce token usage.
    """
    try:
        import pypdf
        from PIL import Image
        import fitz  # PyMuPDF

        doc = fitz.open(stream=pdf_bytes, filetype="pdf")
        images = []

        for page_num in range(min(len(doc), MAX_PAGES)):
            page = doc[page_num]
            
            # Render at 150 DPI — enough for text, 
            # not too large for API
            mat = fitz.Matrix(150/72, 150/72)
            pix = page.get_pixmap(matrix=mat)
            
            img_data = pix.tobytes("png")
            
            # Resize to max 800px width to minimize 
            # token cost while keeping text readable
            img = Image.open(io.BytesIO(img_data))
            if img.width > 800:
                ratio = 800 / img.width
                new_height = int(img.height * ratio)
                img = img.resize(
                    (800, new_height), 
                    Image.LANCZOS
                )
            
            # Convert back to bytes
            buffer = io.BytesIO()
            img.save(buffer, format="PNG", optimize=True)
            buffer.seek(0)
            
            b64 = base64.b64encode(
                buffer.read()
            ).decode("utf-8")
            images.append(b64)

        doc.close()
        return images

    except Exception as e:
        print(f"PDF conversion error: {e}")
        return []

async def extract_drugs_with_openai(
    image_base64: str
) -> Optional[dict]:
    """
    Use GPT-4o-mini vision to extract drug names.
    Cost optimized: low detail mode, minimal tokens.
    Estimated cost per call: ~$0.001-0.002
    """
    if not settings.OPENAI_API_KEY:
        return None

    try:
        async with httpx.AsyncClient(timeout=30) as client:
            response = await client.post(
                "https://api.openai.com/v1/chat/completions",
                headers={
                    "Authorization": 
                        f"Bearer {settings.OPENAI_API_KEY}",
                    "Content-Type": "application/json"
                },
                json={
                    "model": "gpt-4o-mini",
                    "messages": [
                        {
                            "role": "system",
                            "content": SYSTEM_PROMPT
                        },
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "image_url",
                                    "image_url": {
                                        "url": f"data:image/png;base64,{image_base64}",
                                        # LOW detail = 
                                        # ~85 tokens per image
                                        # vs high = 1000+ tokens
                                        "detail": IMAGE_DETAIL
                                    }
                                },
                                {
                                    "type": "text",
                                    "text": "Extract all medication names from this prescription. Return only JSON."
                                }
                            ]
                        }
                    ],
                    # Keep tokens minimal — 
                    # just need a drug list
                    "max_tokens": MAX_TOKENS,
                    "temperature": 0.0  # Deterministic output
                }
            )

            if response.status_code != 200:
                print(f"OpenAI OCR error: {response.text}")
                return None

            data = response.json()
            content = data["choices"][0]["message"]["content"]

            # Log estimated cost
            tokens_used = data.get("usage", {})
            input_tokens = tokens_used.get(
                "prompt_tokens", 0
            )
            output_tokens = tokens_used.get(
                "completion_tokens", 0
            )
            # GPT-4o-mini pricing:
            # $0.00015/1K input, $0.0006/1K output
            cost = (input_tokens * 0.00015 / 1000 + 
                   output_tokens * 0.0006 / 1000)
            print(f"OCR cost: ${cost:.5f} "
                  f"({input_tokens} in, "
                  f"{output_tokens} out tokens)")

            # Parse JSON response
            cleaned = content.strip()
            if "```" in cleaned:
                parts = cleaned.split("```")
                for part in parts:
                    if "{" in part:
                        cleaned = part.replace(
                            "json", ""
                        ).strip()
                        break

            result = json.loads(cleaned)
            return result

    except json.JSONDecodeError as e:
        print(f"OpenAI OCR JSON parse error: {e}")
        print(f"Raw response: {content}")
        return None
    except Exception as e:
        print(f"OpenAI OCR error: {e}")
        return None

def extract_drugs_with_patterns(text: str) -> list[str]:
    """
    Fallback: Pattern-based drug extraction from raw text.
    Used when both OpenAI and Tesseract are unavailable.
    """
    KNOWN_DRUGS = [
        "warfarin", "aspirin", "metformin", "lisinopril",
        "metoprolol", "simvastatin", "atorvastatin",
        "amlodipine", "omeprazole", "amiodarone", "digoxin",
        "fluconazole", "ibuprofen", "clopidogrel", "lithium",
        "phenytoin", "methotrexate", "furosemide",
        "levothyroxine", "sertraline", "fluoxetine",
        "escitalopram", "alprazolam", "lorazepam", "diazepam",
        "tramadol", "gabapentin", "pregabalin", "losartan",
        "valsartan", "ramipril", "bisoprolol", "carvedilol",
        "rosuvastatin", "pravastatin", "pantoprazole",
        "ciprofloxacin", "azithromycin", "amoxicillin",
        "doxycycline", "metronidazole", "clarithromycin",
        "insulin", "glipizide", "sitagliptin", "rivaroxaban",
        "apixaban", "dabigatran", "atenolol", "nifedipine",
        "acetaminophen", "naproxen", "celecoxib", "prednisone",
        "allopurinol", "colchicine", "hydroxychloroquine",
        "spironolactone", "hydrochlorothiazide"
    ]

    ALIASES = {
        "advil": "ibuprofen", "motrin": "ibuprofen",
        "tylenol": "acetaminophen", "coumadin": "warfarin",
        "glucophage": "metformin", "lipitor": "atorvastatin",
        "crestor": "rosuvastatin", "norvasc": "amlodipine",
        "lopressor": "metoprolol", "prinivil": "lisinopril",
        "zestril": "lisinopril", "plavix": "clopidogrel",
        "lasix": "furosemide", "synthroid": "levothyroxine",
        "zoloft": "sertraline", "prozac": "fluoxetine",
        "nexium": "esomeprazole", "prilosec": "omeprazole",
        "zithromax": "azithromycin"
    }

    text_lower = text.lower()
    found = set()

    for drug in KNOWN_DRUGS:
        import re
        pattern = r'\b' + re.escape(drug) + r'\b'
        if re.search(pattern, text_lower):
            found.add(drug)

    for alias, generic in ALIASES.items():
        import re
        pattern = r'\b' + re.escape(alias) + r'\b'
        if re.search(pattern, text_lower):
            found.add(generic)

    return list(found)

async def extract_drugs_from_pdf(
    pdf_bytes: bytes
) -> dict:
    """
    Main OCR pipeline:
    1. Convert PDF to images
    2. Try OpenAI Vision (primary — most accurate)
    3. Signal frontend to use Tesseract (fallback)
    4. Pattern matching on any available text
    """
    result = {
        "drugs_found": [],
        "confidence": "low",
        "pages_processed": 0,
        "method_used": "none",
        "use_tesseract_fallback": False,
        "notes": ""
    }

    # Step 1: Convert PDF to images
    images = await pdf_to_images_base64(pdf_bytes)

    if not images:
        result["notes"] = (
            "Could not process PDF. "
            "Please ensure the file is not corrupted."
        )
        result["use_tesseract_fallback"] = True
        return result

    result["pages_processed"] = len(images)

    # Step 2: Try OpenAI Vision (primary)
    if settings.OPENAI_API_KEY:
        all_drugs = set()
        highest_confidence = "low"

        for i, img_b64 in enumerate(images):
            print(f"Processing page {i+1} with OpenAI...")
            ai_result = await extract_drugs_with_openai(
                img_b64
            )

            if ai_result and ai_result.get("drugs"):
                drugs = [
                    d.lower().strip() 
                    for d in ai_result["drugs"]
                    if d and len(d) > 2
                ]
                all_drugs.update(drugs)

                conf = ai_result.get("confidence", "low")
                if conf == "high":
                    highest_confidence = "high"
                elif (conf == "medium" and 
                      highest_confidence == "low"):
                    highest_confidence = "medium"

                if ai_result.get("notes"):
                    result["notes"] = ai_result["notes"]

        if all_drugs:
            result["drugs_found"] = list(all_drugs)
            result["confidence"] = highest_confidence
            result["method_used"] = "openai_vision"
            return result

        print("OpenAI returned no drugs, "
              "flagging for Tesseract fallback")

    # Step 3: Signal frontend to use Tesseract
    # Send back the images as base64 for browser processing
    result["use_tesseract_fallback"] = True
    result["images_base64"] = images  
    result["method_used"] = "tesseract_fallback"
    result["notes"] = (
        "OpenAI unavailable — using browser OCR"
    )
    return result

File: frontend/src/services/ocrService.js
import * as Tesseract from "tesseract.js";
import * as pdfjsLib from "pdfjs-dist";
import api from "./api";

export async function scanPrescriptionWithBackend(
    file,
    onProgress,
    onStatusChange
) {
    try {
        onStatusChange("Uploading prescription...");
        onProgress(10);

        const formData = new FormData();
        formData.append("file", file);

        onStatusChange("Analyzing with AI...");
        onProgress(30);

        const response = await api.post(
            "/api/ocr/scan-prescription",
            formData,
            {
                headers: { "Content-Type": "multipart/form-data" },
                timeout: 60000
            }
        );

        const data = response.data;
        onProgress(70);

        // If backend says use Tesseract fallback
        if (data.use_tesseract_fallback) {
            onStatusChange("Switching to browser OCR...");

            // If backend sent back images, use them
            if (data.images_base64 && data.images_base64.length > 0) {
                let allText = "";
                const images = data.images_base64;

                for (let i = 0; i < images.length; i++) {
                    onStatusChange(
                        `Browser OCR: page ${i + 1} of ${images.length}...`
                    );
                    const imageUrl =
                        `data:image/png;base64,${images[i]}`;
                    const text = await extractTextFromImage(
                        imageUrl,
                        (p) => {
                            const base = 70 + (i / images.length) * 25;
                            const step = (1 / images.length) * 25;
                            onProgress(
                                Math.round(base + (p / 100) * step)
                            );
                        }
                    );
                    allText += text + "\n";
                }

                const drugs = extractDrugNames(allText);
                onProgress(100);
                onStatusChange("Done!");

                return {
                    success: true,
                    drugs_found: drugs,
                    confidence: drugs.length > 0 ? "medium" : "low",
                    method_used: "tesseract_browser",
                    pages_processed: images.length,
                    raw_text: allText
                };
            } else {
                // No images from backend, process PDF locally
                return await processPrescriptionPDF(
                    file, onProgress, onStatusChange
                );
            }
        }

        // OpenAI succeeded
        onProgress(100);
        onStatusChange("Done!");

        return {
            success: true,
            drugs_found: data.drugs_found || [],
            confidence: data.confidence || "medium",
            method_used: data.method_used,
            pages_processed: data.pages_processed,
            notes: data.notes || "",
            raw_text: ""
        };

    } catch (err) {
        console.error("Backend OCR failed:", err);
        onStatusChange("Trying browser OCR...");

        // Full fallback to local Tesseract
        return await processPrescriptionPDF(
            file, onProgress, onStatusChange
        );
    }
}

// Set PDF.js worker
pdfjsLib.GlobalWorkerOptions.workerSrc = window.pdfjsWorkerSrc ||
    `https://unpkg.com/pdfjs-dist@${pdfjsLib.version}/build/pdf.worker.min.mjs`;

// Common drug name patterns to extract from OCR text
const DRUG_PATTERNS = [
    // Generic names — common drugs
    "warfarin", "aspirin", "metformin", "lisinopril",
    "metoprolol", "simvastatin", "atorvastatin",
    "amlodipine", "omeprazole", "amiodarone", "digoxin",
    "fluconazole", "ibuprofen", "clopidogrel", "lithium",
    "phenytoin", "methotrexate", "furosemide", "spironolactone",
    "levothyroxine", "sertraline", "fluoxetine", "escitalopram",
    "alprazolam", "lorazepam", "diazepam", "tramadol",
    "codeine", "gabapentin", "pregabalin", "losartan",
    "valsartan", "ramipril", "bisoprolol", "carvedilol",
    "rosuvastatin", "pravastatin", "pantoprazole", "esomeprazole",
    "ciprofloxacin", "azithromycin", "amoxicillin", "doxycycline",
    "metronidazole", "clarithromycin", "insulin", "glipizide",
    "sitagliptin", "empagliflozin", "dapagliflozin",
    "rivaroxaban", "apixaban", "dabigatran", "clopidogrel",
    "atenolol", "nifedipine", "verapamil", "diltiazem",
    "hydrochlorothiazide", "chlorthalidone", "acetaminophen",
    "naproxen", "celecoxib", "prednisone", "prednisolone",
    "allopurinol", "colchicine", "hydroxychloroquine",
    "adalimumab", "methotrexate", "sulfasalazine",

    // Brand names mapped to generic
    "advil", "motrin", "tylenol", "coumadin", "glucophage",
    "lipitor", "crestor", "norvasc", "lopressor", "prinivil",
    "zestril", "plavix", "lasix", "synthroid", "zoloft",
    "prozac", "lexapro", "xanax", "ativan", "valium",
    "nexium", "prilosec", "prevacid", "zithromax", "augmentin"
];

// Drug aliases (brand → generic)
const DRUG_ALIASES = {
    "advil": "ibuprofen",
    "motrin": "ibuprofen",
    "tylenol": "acetaminophen",
    "coumadin": "warfarin",
    "glucophage": "metformin",
    "lipitor": "atorvastatin",
    "crestor": "rosuvastatin",
    "norvasc": "amlodipine",
    "lopressor": "metoprolol",
    "prinivil": "lisinopril",
    "zestril": "lisinopril",
    "plavix": "clopidogrel",
    "lasix": "furosemide",
    "synthroid": "levothyroxine",
    "zoloft": "sertraline",
    "prozac": "fluoxetine",
    "lexapro": "escitalopram",
    "xanax": "alprazolam",
    "ativan": "lorazepam",
    "valium": "diazepam",
    "nexium": "esomeprazole",
    "prilosec": "omeprazole",
    "prevacid": "lansoprazole",
    "zithromax": "azithromycin",
    "augmentin": "amoxicillin"
};

export async function convertPdfToImages(file) {
    const arrayBuffer = await file.arrayBuffer();
    const pdf = await pdfjsLib.getDocument({ data: arrayBuffer }).promise;
    const images = [];

    for (let i = 1; i <= Math.min(pdf.numPages, 3); i++) {
        const page = await pdf.getPage(i);
        const viewport = page.getViewport({ scale: 2.0 });

        const canvas = document.createElement("canvas");
        canvas.width = viewport.width;
        canvas.height = viewport.height;

        const ctx = canvas.getContext("2d");
        await page.render({
            canvasContext: ctx,
            viewport
        }).promise;

        images.push(canvas.toDataURL("image/png"));
    }

    return images;
}

export async function extractTextFromImage(
    imageDataUrl,
    onProgress
) {
    const result = await Tesseract.recognize(
        imageDataUrl,
        "eng",
        {
            logger: (m) => {
                if (m.status === "recognizing text" && onProgress) {
                    onProgress(Math.round(m.progress * 100));
                }
            }
        }
    );
    return result.data.text;
}

export function extractDrugNames(rawText) {
    if (!rawText) return [];

    const text = rawText.toLowerCase();
    const lines = text
        .split("\n")
        .map(l => l.trim())
        .filter(l => l.length > 0);

    const foundDrugs = new Map();
    // Map: generic_name → { source, confidence }

    const KNOWN_DRUGS = [
        "warfarin", "aspirin", "metformin", "lisinopril",
        "metoprolol", "simvastatin", "atorvastatin",
        "amlodipine", "omeprazole", "amiodarone", "digoxin",
        "fluconazole", "ibuprofen", "clopidogrel", "lithium",
        "phenytoin", "methotrexate", "furosemide",
        "spironolactone", "levothyroxine", "sertraline",
        "fluoxetine", "escitalopram", "alprazolam", "lorazepam",
        "diazepam", "tramadol", "codeine", "gabapentin",
        "pregabalin", "losartan", "valsartan", "ramipril",
        "bisoprolol", "carvedilol", "rosuvastatin", "pravastatin",
        "pantoprazole", "esomeprazole", "ciprofloxacin",
        "azithromycin", "amoxicillin", "doxycycline",
        "metronidazole", "clarithromycin", "insulin", "glipizide",
        "sitagliptin", "rivaroxaban", "apixaban", "dabigatran",
        "atenolol", "nifedipine", "verapamil", "diltiazem",
        "acetaminophen", "naproxen", "celecoxib", "prednisone",
        "prednisolone", "allopurinol", "colchicine",
        "hydroxychloroquine", "hydrochlorothiazide",
        "empagliflozin", "dapagliflozin"
    ];

    const DRUG_ALIASES = {
        "advil": "ibuprofen",
        "motrin": "ibuprofen",
        "tylenol": "acetaminophen",
        "coumadin": "warfarin",
        "glucophage": "metformin",
        "lipitor": "atorvastatin",
        "crestor": "rosuvastatin",
        "norvasc": "amlodipine",
        "lopressor": "metoprolol",
        "prinivil": "lisinopril",
        "zestril": "lisinopril",
        "plavix": "clopidogrel",
        "lasix": "furosemide",
        "synthroid": "levothyroxine",
        "zoloft": "sertraline",
        "prozac": "fluoxetine",
        "lexapro": "escitalopram",
        "xanax": "alprazolam",
        "ativan": "lorazepam",
        "valium": "diazepam",
        "nexium": "esomeprazole",
        "prilosec": "omeprazole",
        "prevacid": "lansoprazole",
        "zithromax": "azithromycin",
        "augmentin": "amoxicillin",
        "deltasone": "prednisone",
        "medrol": "methylprednisolone",
        "diflucan": "fluconazole",
        "neurontin": "gabapentin",
        "lyrica": "pregabalin",
        "cozaar": "losartan",
        "diovan": "valsartan",
        "coreg": "carvedilol",
        "toprol": "metoprolol",
        "altace": "ramipril",
        "lanoxin": "digoxin"
    };

    // STEP 1 — Extract words per line with strict 
    // word boundary matching only
    // This prevents "verapamil" matching "erapamil" 
    // from garbled OCR text

    for (const line of lines) {
        // Clean line — remove numbers, dots, special chars
        // but keep letters and spaces
        const cleanLine = line
            .replace(/^\d+[\.\)\-\s]+/, "") // Remove "1." "2)" etc
            .replace(/[^a-z\s]/g, " ")
            .replace(/\s+/g, " ")
            .trim();

        if (cleanLine.length < 3) continue;

        // Split into individual words
        const words = cleanLine.split(" ").filter(w => w.length >= 3);

        for (const word of words) {
            // Check aliases first (exact word match)
            if (DRUG_ALIASES[word]) {
                const generic = DRUG_ALIASES[word];
                if (!foundDrugs.has(generic)) {
                    foundDrugs.set(generic, {
                        source: `alias:${word}`,
                        confidence: "high"
                    });
                }
                continue;
            }

            // Check exact drug name match
            if (KNOWN_DRUGS.includes(word)) {
                if (!foundDrugs.has(word)) {
                    foundDrugs.set(word, {
                        source: "exact",
                        confidence: "high"
                    });
                }
                continue;
            }

            // Check if word STARTS WITH a known drug name
            // (handles OCR adding extra chars at end)
            // Minimum 5 chars to avoid false positives
            if (word.length >= 5) {
                for (const drug of KNOWN_DRUGS) {
                    if (drug.length >= 5 && word.startsWith(drug)) {
                        if (!foundDrugs.has(drug)) {
                            foundDrugs.set(drug, {
                                source: "prefix",
                                confidence: "medium"
                            });
                        }
                        break;
                    }
                }
                // Check alias prefix too
                for (const [alias, generic] of
                    Object.entries(DRUG_ALIASES)) {
                    if (alias.length >= 5 && word.startsWith(alias)) {
                        if (!foundDrugs.has(generic)) {
                            foundDrugs.set(generic, {
                                source: `alias_prefix:${alias}`,
                                confidence: "medium"
                            });
                        }
                        break;
                    }
                }
            }
        }
    }

    // STEP 2 — Look for numbered list pattern specifically
    // "1.aspirin", "2.advil" etc — very high confidence
    const numberedPattern = /^\d+[\.\)]\s*([a-zA-Z]+)/gm;
    let match;
    const textCopy = rawText.toLowerCase();

    while ((match = numberedPattern.exec(textCopy)) !== null) {
        const word = match[1].toLowerCase().trim();

        // Exact alias match
        if (DRUG_ALIASES[word]) {
            const generic = DRUG_ALIASES[word];
            foundDrugs.set(generic, {
                source: "numbered_list_alias",
                confidence: "high"
            });
            continue;
        }

        // Exact drug match
        if (KNOWN_DRUGS.includes(word)) {
            foundDrugs.set(word, {
                source: "numbered_list",
                confidence: "high"
            });
            continue;
        }

        // Partial match for numbered lists 
        // (higher tolerance since format is clear)
        for (const drug of KNOWN_DRUGS) {
            if (drug.length >= 4 && (
                word.startsWith(drug) ||
                drug.startsWith(word.slice(0, Math.min(word.length, 8)))
            )) {
                foundDrugs.set(drug, {
                    source: "numbered_list_partial",
                    confidence: "medium"
                });
                break;
            }
        }
        for (const [alias, generic] of
            Object.entries(DRUG_ALIASES)) {
            if (alias.length >= 4 && word.startsWith(alias)) {
                foundDrugs.set(generic, {
                    source: "numbered_list_alias_partial",
                    confidence: "medium"
                });
                break;
            }
        }
    }

    // STEP 3 — Deduplication
    // If both a brand name AND its generic were found,
    // keep only the generic (remove the brand)
    // This prevents "aspirin" + "aspirin" duplicates
    const genericValues = new Set(
        Object.values(DRUG_ALIASES)
    );

    const finalDrugs = [];
    for (const [drug, meta] of foundDrugs.entries()) {
        finalDrugs.push({
            name: drug,
            confidence: meta.confidence,
            source: meta.source
        });
    }

    // Sort: high confidence first, then alphabetical
    finalDrugs.sort((a, b) => {
        const confOrder = { high: 0, medium: 1, low: 2 };
        const confDiff = confOrder[a.confidence] -
            confOrder[b.confidence];
        if (confDiff !== 0) return confDiff;
        return a.name.localeCompare(b.name);
    });

    return finalDrugs.map(d => d.name);
}

export async function processPrescriptionPDF(
    file,
    onProgress,
    onStatusChange
) {
    try {
        onStatusChange("Converting PDF to image...");
        onProgress(5);

        const images = await convertPdfToImages(file);

        if (images.length === 0) {
            throw new Error("Could not extract pages from PDF");
        }

        onStatusChange(`Processing ${images.length} page(s)...`);
        onProgress(20);

        let allText = "";

        for (let i = 0; i < images.length; i++) {
            onStatusChange(
                `Reading page ${i + 1} of ${images.length}...`
            );
            const text = await extractTextFromImage(
                images[i],
                (p) => {
                    const base = 20 + (i / images.length) * 60;
                    const step = (1 / images.length) * 60;
                    onProgress(Math.round(base + (p / 100) * step));
                }
            );
            allText += text + "\n";
        }

        onStatusChange("Extracting medication names...");
        onProgress(85);

        const drugs = extractDrugNames(allText);

        onProgress(100);
        onStatusChange("Done!");

        return {
            success: true,
            drugs_found: drugs,
            raw_text: allText,
            pages_processed: images.length,
            confidence: drugs.length > 0 ? "medium" : "low"
        };

    } catch (err) {
        return {
            success: false,
            error: err.message,
            drugs_found: [],
            raw_text: ""
        };
    }
}
